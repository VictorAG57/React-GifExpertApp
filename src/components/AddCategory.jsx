import React, { useState } from 'react';
import PropsTypes from "prop-types";

/** Función del componente AddCategory
 * 
 * Esté me permite agregar una categoria a mi página, el usuario ingresa
 * en una caja de texto la categoría que desea agragar, y al precionar enter (submit)
 * actualiza el estado y lo agrega a una lista con las demás categorias.
 * 
 * Esté componente complementa al componente pruncipal GifExpertApp.
 */

export const AddCategory = ({ setCategory }) => {

    const [inputValue, setInputValue] = useState("")

    //Obtenemos el valor de nuestra caja de texto para actualizar el estado.
    //Lo escrito por el usuario
    const handleInputChange = (e)=> {
        setInputValue(e.target.value)
        console.log("handleInputChange llamado")
    };

    //Cambiamos el estado con "handleSubmit", agregamos un elemento más al etsado
    const handleSubmit = (e)=> {
        //Evitamos que el usuario al dar enter refresque la página
        //ya que es un formulario
        e.preventDefault()

        //Meidante el if, evitamos que se pueda agregar a la lista un elemento
        //sin coontenido
        if(inputValue.trim().length > 2){
            //Primero colocamos inputValue, para cuando se renderize, lo
            //primero que aparezac sea la ultima busqueda y no aparezca hasta el final
            setCategory( cats => [ inputValue, ...cats])
            setInputValue("")
            console.log("setCategory llamado")
        }
    };

    return (
        //Usamos el onSubmit pra implementar la función "handleSubmit"
        <form onSubmit={ handleSubmit }>
            <input type="text" 
                value={ inputValue } 
                onChange={ handleInputChange }>
            </input>
        </form>
    );
}

//Obligamos a que "setCategory" sea obligatoria
AddCategory.propTypes = {
    setCategory: PropsTypes.func.isRequired
};

