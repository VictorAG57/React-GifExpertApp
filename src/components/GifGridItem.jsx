import React from 'react';
import "../index.css";
import PropsTypes from "prop-types";
import { prototype } from 'webpack-sources/lib/CachedSource';


/**
 * EstÃ© componente es el encargado de la estructura de cada
 * carta de gif que aparece.
 */

export const GifGridItem = ({ id, title, url}) => {
    
    // console.log(img)

    return (
        <div className="card animate__animated animate__bounce animate__delay-2s animate_fadeIn">
            <img src={ url } alt={ title }></img>
            <p>{ title }</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropsTypes.string.isRequired,
    url: PropsTypes.string.isRequired
}