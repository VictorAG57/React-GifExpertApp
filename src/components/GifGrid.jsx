import React, { Fragment} from 'react';
import PropTypes from "prop-types";

import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from "./GifGridItem";

//Se encarga de la estructura de cada Item de gif, para acomodarla
//en una estructutura, los organiza todos, para enviar la infotmación
//a GifExpertApp y la renderize.

//Resive omo argumento la categoría enviada por el usuario para que useFetchGifs
//se encargue de estructurar la información, que as u vez se conecta con nuestro helper
//para realizar la petición.

export const GifGrid = ({ category }) => {

    //useFetchGifs me traera la información de mis gifs
    const { data:images, loading } = useFetchGifs( category )

    return (
        <Fragment>
            <h3 className="animate__animated animate__bounce animate__delay-2s">{ category }</h3>

            { loading && <p className="animate__fadeInRight" >Loading</p> }

            <div className="card-grid">
                { //También puedes extraer las propiedades en img --> {id, title}
                //Manda como parametros la url:img, y el titulo:id a nuestro <GifGridItem>
                images.map( img => ( 
                    <GifGridItem
                        key={img.id} 
                        {...img}>
                    </GifGridItem> 
                    ))
                }
            </div> 
           
        </Fragment>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

// const [images, setImages] = useState([])

    //Usaos useEffect como acciones secundarias, y aqui la usamos con el fin
    //de que una vez cargue la página realiza la petición http para traer el contenido.
    
    // useEffect( ()=> {
    //     GetGifs( category )
    //         .then(imgs => setImages( imgs ))
    // }, [ category ]) 

    //Colocamos [category] como argumento, por si es 
    //necesario que useEffect se vuelva a renderizar, aunque no vaya a pazar pero
    //así eliminamos un warning que nos aparecera en cosola.

 
    /*
        <ol>
            { //También puedes extraer las propiedades en img --> {id, title}
                images.map( img => (<li key={ img.id }> { img.title } </li>) )
            }
        </ol>
    */