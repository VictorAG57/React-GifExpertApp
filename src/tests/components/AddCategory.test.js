import { shallow } from "enzyme"
import { AddCategory } from "../../components/AddCategory"
import "@testing-library/jest-dom"

describe('Debe de renderizar correctamente el componente', () => {

    // Funciona igual que la línea de abajo const setCategory = ()=> {}
    // Solamente que jest.fn, funcionara como si fuera una función,
    //pero nos dejara ver mas caracteristicas.
    const setCategory = jest.fn()
    let wrapper = shallow(<AddCategory setCategory={setCategory}></AddCategory>)

    //Usamos el beforeEach para dar un ciclo de vida, y cuando tengas que realizar
    //alguna tarea en repetidas ocasiones el valor no se altere.
    //Por ejemplo aquí se ejecuta la funcion set Category al tener un argumento por priemra vez
    //pero al colocar el wrapper dentro, limpiamos el formulario y así no se ejecuta la función
    // de manera predeterminada.
    beforeEach(()=> {
        //Se usa cuando tenemos algún mock o usa simulación, s elimpie
        jest.clearAllMocks()
        wrapper = shallow(<AddCategory setCategory={setCategory}></AddCategory>)
    })
    
    test('Debe de mostrarse correctamente el componente ', () => {

        expect(wrapper).toMatchSnapshot()
        
    });

    test('Debe de cambiar la caja de texto', () => {

        const input = wrapper.find("input")
        const value = "Hola mundo"

        //Simulamos un cambio en el input, y colocamos el { target: {value}}
        //porque si no nos marcaria error, ya que ay un parametro en la función (e)
        //y colocamos target porque nos ineresa referenciar al evento (change) y al valor
        input.simulate("change", { target: { value }})  
    });

    test('No debe postear la información con submit', () => {

        //Aqui simulamos que en nuestro componente, al dar submit, no 
        //se ejecute, el argumento que resive nuestro simulate aparte de submit
        // es prevent default, que es la función que por defecto evita que el submit se 
        //ejecute, y esa misma función es la que progarmamos en nuestra logica de AddCategory 

        wrapper.find("form").simulate("submit", { preventDefault(){} })
        
        //.not --> no, no suceda, no ocurra
        //.toHaveBeenCalled --> que se alla llamado
        //.not.toHaveBeenCalled --> que no se alla llamado
        //.toHaveBeenCalledWith --> que se alla llamado con argumentos/función
        //.toHaveBeenCalledTimes(vecesLlamadas) --> que se alla llamado cierta cantidad de veces


        //Evaluamos que al dar submit, setCategory no se allá llamado, por ello 
        // usamos .not.toHaveBeenCalled.
        expect( setCategory ).not.toHaveBeenCalled()
    });
    
    test('Debe de llamar el setCategory y limpiar la caja de texto ', () => {
        
        const input = wrapper.find("input")
        const value = "Hola mundo"

        // 1. Simular inputChange, simulamos un cambio en la caja de texto
        input.simulate("change", { target: { value }})

        // 2. Simular el submit, simulamos al dar submit, se ejecute preventDefault()
        wrapper.find("form").simulate("submit", { preventDefault(){} })

        // 3. Simular setCategory, simulamos que se allá llamado
        expect( setCategory ).toHaveBeenCalled()

        // 4. Simulamos que el valor de input se alla reseteado a ""
        expect( input.prop("value") ).toBe("")
    }); 
    
});
