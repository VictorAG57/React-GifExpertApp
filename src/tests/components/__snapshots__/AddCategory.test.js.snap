// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debe de renderizar correctamente el componente Debe de mostrarse correctamente el componente  1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
